pkg/imageallowrules/imageallowrules.go

```diff
@@ -123,21 +162,84 @@ func CheckImageAgainstRules(ctx context.Context, c client.Reader, namespace stri
 					}
 				}
 				if !anyOfOK {
-					notAllowedErr.SubrulePath += ".anyOf"
 					if len(anyOfErrs) == len(rule.SignedBy.AnyOf) {
 						// we had errors for all anyOf rules (not failed verification, but actual errors)
 						e := fmt.Errorf("error verifying image %s against %s/%s.signatures.anyOf.*: %w", image, imageAllowRule.Namespace, imageAllowRule.Name, merr.NewErrors(anyOfErrs...))
 						logrus.Errorln(e.Error())
-						return e
 					}
-					logrus.Warnf("image %s is not allowed as per %s/%s.signatures.anyOf", image, imageAllowRule.Namespace, imageAllowRule.Name)
-					return notAllowedErr
+					continue iarLoop // failed or errored in all anyOf, try next IAR
 				}
 			}
 		}
+
+		return nil
+	}
+
+	return &ErrImageNotAllowed{Image: image}
+}
+
+func imageCovered(image name.Reference, digest name.Digest, iar v1.ImageAllowRuleInstance) bool {
+	for _, pattern := range iar.Images {
+		// empty pattern? skip (should've been caught by IAR validation already)
+		if strings.TrimSpace(pattern) == "" {
+			continue
+		}
+
+		// not a pattern and image name is not an exact match? skip
+		if !imagepattern.IsImagePattern(pattern) {
+			if image.Name() != pattern && (digest.Identifier() != "" && digest.Name() != pattern) {
+				continue
+			}
+		}
+
+		parts := strings.Split(pattern, ":")
+		contextPattern := parts[0]
+		tagPattern := ""
+		if len(parts) > 1 && !strings.Contains(parts[len(parts)-1], "/") {
+			tagPattern = parts[len(parts)-1]
+		}
+
+		if err := matchContext(contextPattern, image.Context().String()); err != nil {
+			logrus.Debugf("image %s not in scope of ImageAllowRule %s/%s: %v", image, iar.Namespace, iar.Name, err)
+			continue
+		}
+
+		if tagPattern != "" {
+			if err := matchTag(tagPattern, image.Identifier()); err != nil {
+				logrus.Debugf("image %s not in scope of ImageAllowRule %s/%s: %v", image, iar.Namespace, iar.Name, err)
+				continue
+			}
+		}
+
+		return true
+	}
+	return false
+}
+
+// matchContext matches the image context against the context pattern, similar to globbing
+func matchContext(contextPattern string, imageContext string) error {
+	re, _, err := imagepattern.NewMatcher(contextPattern)
+	if err != nil {
+		return fmt.Errorf("error parsing context pattern %s: %w", contextPattern, err)
+	}
+
+	if re.MatchString(imageContext) {
+		return nil
+	}
+
+	return fmt.Errorf("image context %s does not match pattern %s (regex: `%s`)", imageContext, contextPattern, re.String())
```

Comment:
It's a fair point though. It's odd and I just did it, because the calling code doesn't care why it cannot use it, whether it doesn't match or failed to evaluate.